### DISCUSSION TITLE
Unexpected failover behavior in karmada

### DISCUSSION THREAD

**Discussion by NickYadance** [karmada-io/karmada#5309] [Q&A] [answered]
> Releated issue: https://github.com/karmada-io/karmada/issues/5150

Karmada deleted part of resources in a member cluster when the member cluster is down for about ~2minutes, at the same time, Karmada creates all resources again after it recovers.

I try to explain this with failover and reproduce it, but the failover and graceful eviction in Karmada requires that:

1. The cluster down time >failoverTimeout, then `NoExecute` taint can be added
2. The cluster down time >tolerationSeconds , then graceful eviction task can be added into resourcebindings
3. The graceful eviction task is done, so that orphan works will be deleted

Karmada needs to wait at least failoverTimeout(5min) + tolerationSeconds(5min) before failover actually happens, which doesn't match the actual recovery time of 2min. 

After searching for quite a while i still cannot find a possible reasonable cause for this, any suggestions ?

**Comment by NickYadance** [comment, marked as answer]
Found a possible explanation for this related to `getAPIEnablements` returning partial results.
https://github.com/karmada-io/karmada/blob/b4b6d692ca963c3d53de4a61077f4380613be0a5/pkg/controllers/status/cluster_status_controller.go#L440-L471

In my case the cluster doesn't shutdown right away but in a graceful way, thus reject incoming connections then close existing connections. It's possible for getAPIEnablements to get partial results during the process, which includes only parts of the GVK, and set them into cluster status. **This equals to removing GVK from the cluster manually.**

After the NoSchedule taint is added to the cluster, scheduler will immediatily reschedule all resource bindings binded to the cluster. 
The ApiEnablement plugin in scheduler will fail due to missing GVK in cluster status. So scheduler will remove the unhealthy cluster from the resource bindings of these missing GVKs, producing BindingUpdate.

Now the issue happens, because the unhealthy cluster no longer exists in resource binding, all the existing works will be deleted, and all workloads will be deleted after the cluster is back alive.  This also explains why there are only part of the resources being deleted in my case.

Maybe worthy a look @XiShanYongYe-Chang 

<img width="1053" alt="image" src="https://github.com/user-attachments/assets/9730c34a-b449-4f1b-9a6b-c9aea20807d5">

<img width="1062" alt="image" src="https://github.com/user-attachments/assets/8e0fbdca-3c7c-4de7-bb60-150d8aff46c4">

**Reply by whitewindmills** [reply]
yes, I have the same problem.
